vignettes/namespaces.Rmd

```diff
@@ -0,0 +1,416 @@
+xml2::xml_add_child(pgp, "md:text", "hello from the md namespace")
+dc
+```
+
+Now we can see that there are three text nodes, one of which has the `md` 
+namespace prefix. If we select the nodes with that prefix and without the 
+prefix, we will get one and two nodes, respectively.
+
+```{r}
+xml2::xml_find_all(dc, "//md:text") # one node
+xml2::xml_find_all(dc, "//text")    # two nodes
+```
+
+If a namespace is defined in the document with a prefix, *only nodes with that
+prefix are considered to be inside the namespace*. This becomes important when
+we want to pass our XML document through a stylesheet that expects the incoming
+nodes to have a specific namespace, which is exactly how we transform the XML
+representation of markdown back to markdown.
+
+## Commonmark
+
+The {tinkr} package streamlines the process of markdown to xml and back again.
+We use `commonmark::markdown_xml()` as a starting point to generate valid XML:
+
+
+```{r cmark}
+cat(cmk <- commonmark::markdown_xml("this is a **test**"))
+xml <- xml2::read_xml(cmk)
+xml
+```
+
+### Commonmark uses a default namespace
+
+You can see from the commonmark output that it has a **default namespace** that
+resolves to `http://commonmark.org/xml/1.0`, which means that we need to use
+the default namespace if we want to munge the data:
+
+```{r munge}
+xml2::xml_find_all(xml, "//d1:text")
+```
+
+### Using a semantic prefix with the default namespace
+
+To make things more semantic, we could rename the namespace to have the "md"
+prefix and carry around that object. Note: an `xml_namespace` object is a named
+character vector, so we can create it with `structure()` and use it to introduce
+semantically sensible XPath queries
+
+```{r}
+ns <- structure(c(md = "http://commonmark.org/xml/1.0"), class = "xml_namespace")
+xml2::xml_find_all(xml, "//md:text", ns)
+```
+
+Of course, now if we want to make any semantic XPath query, we need to include
+both a prefix and a namespace object.
+
+### Transforming XML to markdown with XSLT
+
+The commonmark namespace allows us to transform our document to markdown using 
+an XSLT stylesheet, which is---that's right---an XML document:
+
+```{r stysh}
+sty <- xml2::read_xml(tinkr::stylesheet())
+sty
+```
+
+Each `xsl:template` node in this stylesheet matches against a specific node in
+the commonmark namespace (prefix: `md`) and emits text based on that node. This
+allows us to write back to markdown:
+
+```{r}
+cat(xslt::xml_xslt(xml, sty))
+```
+
+We can in this way programatically transform the content of the markdown. In 
+this example, we can change the `**test**` to be an inline R code chunk that
+emits `_test_`. 
+
+```{r}
+xml <- commonmark::markdown_xml("this is a **test**") %>%
+  xml2::read_xml()
+
+xml2::xml_find_all(xml, "//md:strong", ns) %>%
+  xml2::xml_set_name("code") %>%
+  xml2::xml_set_text("r cat('_test_')")
+
+sty <- xml2::read_xml(tinkr::stylesheet())
+cat(xslt::xml_xslt(xml, sty))
+```
+
+### Perils: adding nodes
```

Comment:
Is "perils" a common word? I understand it because it looks like the French word, but might it be a better idea to use "risks"?

Reply:
I think peril gives a better sense of "something that can not be avoided on this path" as opposed to risk, which has a random component.